src/pm/xbps.rs

```diff
@@ -0,0 +1,241 @@
+    }
+}
+
+#[async_trait]
+impl Pm for Xbps {
+    /// Gets the name of the package manager.
+    fn name(&self) -> &str {
+        "xbps"
+    }
+
+    fn cfg(&self) -> &crate::dispatch::Config {
+        &self.cfg
+    }
+
+    /// Q generates a list of installed packages.
+    async fn q(&self, kws: &[&str], flags: &[&str]) -> Result<()> {
+        if kws.is_empty() {
+            return self
+                .run(Cmd::new(&["xbps-query", "-l"]).kws(kws).flags(flags))
+                .await;
+        }
+
+        let lines: Vec<_> = stream::iter(kws)
+            .map(Ok)
+            .and_then(|&pkg| async {
+                let cmd = Cmd::new(&["xbps-query", "--property", "pkgver", pkg]).flags(flags);
+                self.check_output(cmd, PmMode::Mute, &Strategy::default())
+                    .await
+                    .map_err(|err| (pkg.to_string(), err))
+            })
+            .collect()
+            .await;
+
+        let mut stdout = tokio::io::stdout();
+        for line in lines {
+            match line {
+                Ok(line) => stdout.write_all(&line).await?,
+                Err((pkg, Error::CmdStatusCodeError { .. })) => {
+                    let msg = format!("error: package '{}' was not found\n", pkg);
+                    stdout.write_all(msg.as_bytes()).await?;
+                }
+                Err((_, other)) => return Err(other),
+            };
+        }
+
+        Ok(())
+    }
+
+    /// Qe lists packages installed explicitly (not as dependencies).
+    async fn qe(&self, kws: &[&str], flags: &[&str]) -> Result<()> {
+        if kws.is_empty() {
+            return self
+                .run(Cmd::new(&["xbps-query", "-m"]).kws(kws).flags(flags))
+                .await;
+        }
+
+        let lines: Vec<_> = stream::iter(kws)
+            .filter(|&&pkg| async {
+                let check_cmd =
+                    Cmd::new(&["xbps-query", "--property", "automatic-install", pkg]).flags(flags);
+                let result = self
+                    .check_output(check_cmd, PmMode::Mute, &Strategy::default())
+                    .await;
+
+                // if a package is manually installed then the automatic-install field is empty
+                match result {
+                    Ok(auto_status) => auto_status.is_empty(),
+                    _ => true,
+                }
+            })
+            .map(Ok)
+            .and_then(|&pkg| async {
+                let cmd = Cmd::new(&["xbps-query", "--property", "pkgver", pkg]).flags(flags);
+                self.check_output(cmd, PmMode::Mute, &Strategy::default())
+                    .await
+                    .map_err(|err| (pkg.to_string(), err))
+            })
+            .collect()
+            .await;
+
+        let mut stdout = tokio::io::stdout();
+        for line in lines {
+            match line {
+                Ok(line) => stdout.write_all(&line).await?,
+                Err((pkg, Error::CmdStatusCodeError { .. })) => {
```

Comment:
In terms of error recovery on failure, I think you could be even more precise. I guess `xbps` will give a specific error code when a package is not found, so that your error message will be guaranteed to pop up at the right time (I'm really afraid of giving inappropriate instructions to users @_@). And if not, you can still match against the output :)